store/auth.go

```diff
@@ -166,3 +172,148 @@ func DischargeAuthCaveat(username, password, macaroon, otp string) (string, erro
+		identityURI: cfg.IdentityURI,
+		client: &http.Client{
+			Transport: &LoggedTransport{
+				Transport: http.DefaultTransport,
+				Key:       "SNAPD_DEBUG_HTTP",
+			},
+		},
+	}
+}
+
+func (s *SnapUbuntuStoreAuthService) requestDeviceNonce() ([]byte, error) {
+	const errorPrefix = "cannot authenticate device to store: failed to get nonce: "
+
+	noncesURL, err := s.identityURI.Parse("nonces")
+	if err != nil {
+		return nil, err
+	}
+	req, err := http.NewRequest("POST", noncesURL.String(), nil)
+	if err != nil {
+		return nil, err
+	}
+	req.Header.Set("Accept", "application/json")
+
+	resp, err := s.client.Do(req)
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != 200 {
+		return nil, fmt.Errorf(errorPrefix+"store server returned status %d", resp.StatusCode)
+	}
+
+	dec := json.NewDecoder(resp.Body)
+	var responseData struct {
+		Nonce string `json:"nonce"`
+	}
+	if err := dec.Decode(&responseData); err != nil {
+		return nil, fmt.Errorf(errorPrefix+"%v", err)
+	}
+
+	if responseData.Nonce == "" {
+		return nil, fmt.Errorf(errorPrefix + "no nonce returned")
+	}
+	return []byte(responseData.Nonce), nil
+}
+
+func (s *SnapUbuntuStoreAuthService) requestDeviceMacaroon(serialAssertion []byte, nonce []byte, signature []byte) (string, error) {
+	const errorPrefix = "cannot authenticate device to store: failed to get macaroon: "
+
+	type deviceMacaroonRequest struct {
+		SerialAssertion string `json:"serial-assertion"`
+		Nonce           string `json:"nonce"`
+		Signature       string `json:"signature"`
+	}
+
+	jsonData, err := json.Marshal(deviceMacaroonRequest{
+		SerialAssertion: string(serialAssertion),
+		Nonce:           string(nonce),
+		Signature:       string(signature),
+	})
+
+	sessionsURL, err := s.identityURI.Parse("sessions")
+	if err != nil {
+		return "", err
+	}
+	req, err := http.NewRequest("POST", sessionsURL.String(), bytes.NewBuffer([]byte(jsonData)))
+	if err != nil {
+		return "", err
+	}
+	req.Header.Set("Accept", "application/json")
+	req.Header.Set("Content-Type", "application/json")
+
+	resp, err := s.client.Do(req)
+	if err != nil {
+		return "", err
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != 200 {
+		// TODO: We should return a more useful error message if
+		// possible. The store's device macaroon APIs produce
+		// error objects in a different format from the user
+		// macaroon APIs, so they should probably be assimilated
+		// first.
+		return "", fmt.Errorf(errorPrefix+"store server returned status %d", resp.StatusCode)
+	}
+
+	dec := json.NewDecoder(resp.Body)
+	var responseData struct {
+		Macaroon string `json:"macaroon"`
+	}
+	if err := dec.Decode(&responseData); err != nil {
+		return "", fmt.Errorf(errorPrefix+"%v", err)
+	}
+
+	if responseData.Macaroon == "" {
+		return "", fmt.Errorf(errorPrefix + "no macaroon returned")
+	}
+	return responseData.Macaroon, nil
+}
+
+type signNonceFunc func(deviceKeyEncoded []byte, nonce []byte) ([]byte, error)
```

Comment:
I don't understand the signature of this, I don't see the role of the store package in picking the interface to talk to the gadget at this level of detail,

something like signNonceFunc(nonce []byte) ([]byte, error) would seem enough,

or passing the full serial assertion and nonce if we feels we need a way to address/index the key (instead of assuming a current one is known at the device management level) and the code closer to the gadget deals with the details.

OTH one concern I share that was raised at the sprints is that we don't want ideally to force the gadget to sign arbitrary things to some extent, so some agreement about constraints on the format of the nonce (prefix and length constraints if not full grammar) would be good